Initial code copy from Koans solutions

mattcasiro opened the pull request:
- Copy over unit tests from relevant Koans modules
- Copy over final solutions from Koans modules dice and score

mattcasiro committed:
Create code for calculating score of a single turn

See GREED_RULES.txt for details on how the score is calculated.

Refactored Score based on code review feeback

- Moved operations out of main method to private helper methods

- Re-wrote hash creation to be more verbose using .each instead of
  .inject

- Re-wrote scoring method to use .inject instead of .each with an
  independant score variable, and broke out the triples calculation from
  the case statement to reduce repeated code

mattcasiro committed:
Create code for rolling a given number of dice.

Results are placed in an array and saved under the instance variable
@values.

mattcasiro committed:
Create unit tests for dice.rb

Tests taken from Ruby Koans

mattcasiro committed:
Create unit tests for score.rb.

Tests pulled from Ruby Koans

luukveenis commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,26 @@
+module Greed
+  module Score
+    def calcScore(dice)
```
small syntax thing, but normally we underscore method names in ruby instead of camelCase, ie:
`def calc_score`

luukveenis commented on lib/greed/dice.rb:
```diff
@@ -0,0 +1,10 @@
+
+    # Roll -num- six sided dice and place results in an array
+    def roll(num)
+      @values = (1..num).map { rand(1..6) }
```
It's a little weird to use map here since you're not using the things you're iterating  over, maybe something like:
`Array.new(num) { rand(1..6) }`

Sinetheta commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,26 @@
+module Greed
+  module Score
+    def calcScore(dice)
+      raise ArgumentError.new unless dice
```
What would happen if this line were not here?

mattcasiro replied:
If nil is passed, the program crashes on dice.inject (NoMethodError) - this line is intended to result in more predictable behavior.

Sinetheta commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,26 @@
+      score = 0
+
+      # Create hash of count of each die
+      counts = dice.inject(Hash.new(0)) {|hash, num| hash[num] += 1; hash}
```
Why is `Hash.new(0)` needed here instead of the object literal notation `{}`?

mattcasiro replied:
The hash requires a default value of 0 for the += operation to work correctly. I'm not sure of another way to do this other than as shown above.

Sinetheta replied:
You got it. The alternative would be defining a start value for each key that you need which doesn't yet exist.

Sinetheta commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,26 @@
+      # Create hash of count of each die
+      counts = dice.inject(Hash.new(0)) {|hash, num| hash[num] += 1; hash}
+
+      # Calculate score
```
You've broken the scoring process into steps, did you consider separating those jobs into their own methods?

Sinetheta commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,26 @@
+      counts = dice.inject(Hash.new(0)) {|hash, num| hash[num] += 1; hash}
+
+      # Calculate score
+      counts.each do |die, count|
```
Previously you used `inject` to reduce an array into an object, could it be used here too instead of having a `score` variable?

Sinetheta commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,26 @@
+
+      # Calculate score
+      counts.each do |die, count|
+        trips, remainder = count / 3, count % 3
```
Generally, you want one statement per line. Destructured assignment is great when you are presented with a complex return value, but in this case we're just doing two different things on the same line for the sake of brevity.

http://www.oualline.com/books.free/style/c04.html

Sinetheta commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,26 @@
+      counts.each do |die, count|
+        trips, remainder = count / 3, count % 3
+
+        case die
```
I like this way of tallying score, but it looks really close to a more general solution. If we gave a name to `die * 100` and noticed that the only outlier is `die = 1`, which behaves remarkedly like `die = 10` could you simply this calculation to not require a case statement?

Sinetheta commented on lib/greed/score.rb:
```diff
@@ -1,24 +1,24 @@
 module Greed
-    def calcScore(dice)
```
In the case of a PR which hasn't been merged yet, it's a good idea to rebase these fixups back in time so that they don't

A) make for a noisy file history
B) interfere with other rebasing you might do before merging your PR

Let me know if you want a hand using http://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html

gmacdougall commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,24 @@
+module Greed
+  def calcScore(dice)
+    raise ArgumentError.new unless dice
```
It's not necessary to instantiate the error in Ruby. I would recommend:

``` Ruby
raise ArgumentError, "Dice must be provided" unless dice
```

gmacdougall commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,24 @@
+    score = 0
+
+    # Create hash of count of each die
+    counts = dice.inject(Hash.new(0)) {|hash, num| hash[num] += 1; hash}
```
I would not let this line of code in to a project that I was working on. It's not immediately clear what's happening. See if you can find a way to make it more verbose (through code) rather than relying on a single complicated line.

gmacdougall commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,24 @@
+
+    # Calculate score
+    counts.each do |die, count|
+      trips, remainder = count / 3, count % 3
```
We don't pay by the line for code. I would make this:

``` Ruby
trips = count / 3
remainder = count % 3
```

Murph33 commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,45 @@
+  def calc_score(dice)
+    # Catch nil parameter
+    raise ArgumentError, "Dice must be provided" unless dice
+    countHash = count_dice_results(dice)
```
prefer snake case to camelcase in ruby for a variable like `count_hash`

Murph33 commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,45 @@
+    end
+
+  # Calculate score based on GREED_RULES.txt
+  private
```
redundant private

Murph33 commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,45 @@
+
+  # Create a hash where key = die face and value = number of occurences
+  private
+    def count_dice_results(dice)
```
Just an indentation thing but we don't generally indent beneath private

Murph33 commented on lib/greed/score.rb:
```diff
@@ -0,0 +1,45 @@
+
+        # 100 points for any individual 1's (not in a set of three)
+        # 50 points for any indivudal 5's (not in a set of three)
+        individualScore = case die
```
variables throughout in here are camelcase instead of snake_case

mattcasiro committed:
Refactored from code review

- Removed redundant private flag

- Removed tab-in after private

- Converted variable names to snake-case
